aten/src/ATen/SparseCsrTensorImpl.cpp

```diff
@@ -0,0 +1,153 @@
+                  .device(SparseCsrTensorSetToDeviceType(key_set))
+                  .dtype(data_type)) // values
+      ) {}
+
+SparseCsrTensorImpl::SparseCsrTensorImpl(
+    at::DispatchKeySet key_set,
+    const caffe2::TypeMeta& data_type,
+    at::Tensor crow_indices,
+    at::Tensor col_indices,
+    at::Tensor values)
+    : TensorImpl(key_set, data_type, values.device()),
+      crow_indices_(std::move(crow_indices)),
+      col_indices_(std::move(col_indices)),
+      values_(std::move(values)) {}
+
+void SparseCsrTensorImpl::resize_and_clear_(
+    int64_t nnz_size,
+    IntArrayRef size) {
+  // call crow_indices().options() here since the struct contructor calls the
+  // tensor constructor with args for device specific init.
+  auto empty_crow_indices = at::empty(size[0] + 1, crow_indices().options());
+  auto empty_col_indices = at::empty(nnz_size, col_indices().options());
+  auto empty_values = at::empty(nnz_size, values().options());
+
+  crow_indices_ = empty_crow_indices;
+  col_indices_ = empty_col_indices;
+  values_ = empty_values;
+  sizes_and_strides_.set_sizes(size);
+}
+
+void SparseCsrTensorImpl::resize_as_sparse_csr_tensor_(const Tensor& src) {
+  crow_indices_ = at::empty_like(
+      src.crow_indices(),
+      src.crow_indices().options(),
+      src.crow_indices().suggest_memory_format());
+  col_indices_ = at::empty_like(
+      src.col_indices(),
+      src.col_indices().options(),
+      src.col_indices().suggest_memory_format());
+  values_ = at::empty_like(
+      src.values(),
+      src.values().options(),
+      src.values().suggest_memory_format());
+  sizes_and_strides_.set_sizes(src.sizes());
+}
+
+void SparseCsrTensorImpl::set_member_tensors(
+    const Tensor& crow_indices,
+    const Tensor& col_indices,
+    const Tensor& values) {
+  auto crow_indices_type = crow_indices.scalar_type();
+  auto col_indices_type = col_indices.scalar_type();
+
+  TORCH_CHECK(
+      crow_indices_type == col_indices_type,
+      "both crow_indices and col_indices should have the same type.");
+  TORCH_CHECK(
+      crow_indices_type == kInt || crow_indices_type == kLong,
+      "crow_indices and col_indices must be an int32 or int64 type, but got: ",
+      crow_indices_type);
+  TORCH_CHECK(
+      values.scalar_type() == typeMetaToScalarType(dtype()),
+      "dtype of values (",
+      values.scalar_type(),
+      ") must match dtype of sparse tensor (",
+      typeMetaToScalarType(dtype()),
+      ")");
+
+  TORCH_CHECK(
+      col_indices.layout() == kStrided,
+      "expected col_indices to be a strided tensor, but got indices of layout ",
+      col_indices.layout());
+  TORCH_CHECK(
+      crow_indices.layout() == kStrided,
+      "expected crow_indices to be a strided tensor, but got crow_indices of layout ",
+      crow_indices.layout());
+  TORCH_CHECK(
+      values.layout() == kStrided,
+      "expected values to be a strided tensor, but got values of layout ",
+      values.layout());
```

Comment:
Isn't that the same as strided?

Reply:
@v0dro , no, it is not the same. strided can be non-contiguous.

When using non-contiguous strided tensor as values, then the following code is wrong
https://github.com/pytorch/pytorch/pull/50937/files#diff-5b05f47f16f880c1e4e943963a974aed1e12d0c2d305015864fb2e39630c9aeaR156
for instance.

COO allows non-contiguous indices and values. Following that, all access to crow_indices, indices, and values attributes that use `data_ptr`, should use `contiguous()` method. See, for instance, https://github.com/pytorch/pytorch/blob/b074a24394003ecfe184c81d5ef7daee1a757814/aten/src/ATen/native/sparse/SparseMatMul.cpp#L172

Reply:
Ah I see. Added a check for that.

Reply:
Requiring contiguity here could only be a temporary solution, IMHO.
COO tensors can have non-contiguous indices and values:
```
>>> import torch
>>> a = torch.tensor([[0, 1, 0, 2], [3, 0, 4, 0]])
>>> a.to_sparse()
tensor(indices=tensor([[0, 0, 1, 1],
                       [1, 3, 0, 2]]),
       values=tensor([1, 2, 3, 4]),
       size=(2, 4), nnz=4, layout=torch.sparse_coo)
>>> s = a.to_sparse()
>>> s2 = torch.sparse_coo_tensor(s._indices()[:, ::2], s._values()[::2])
>>> s2
tensor(indices=tensor([[0, 1],
                       [1, 0]]),
       values=tensor([1, 3]),
       size=(2, 2), nnz=2, layout=torch.sparse_coo)
>>> s2._values().is_contiguous()
False
>>> s._values().is_contiguous()
True
```

Why CSR tensors should not have the same property? I think the answer is that it does not make sense to take a slice of `crow_indices` because the results would not be something that could be reused in sparse_csr_tensor without preprocessing, in the COO case we could do that.

However, the `values` and `indices` could be sliced and reused in sparse_csr_tensor, so I think `values` and `indices` should be allowed to be non-contiguous tensors. So, instead of requiring contiguity of `values`, could you rephrase the exception message to as "support for non-contiguous values is not implemented yet", or better yet, just use `._values.().contiguous()` where relevant.

Reply:
tbh I don't think the non-contiguous indices/values in COO tensors are intentionally allowed, people just forgot to add checks for it. It seems pretty unlikely to me that kernels are going to be able to handle noncontiguous indices/values, which to me has me vote to enforce contiguity. I don't think `_values().contiguous()` is better as it is a perf trap.